### PR TITLE
LowPtElectrons: updated BDT model and code base for ID

### PR DESCRIPTION
**Pull request by bainbrid**
__This PR concerns changes to the BDT model used for the identification of LowPtElectrons.__

__This PR requires the following external: cms-data/RecoEgamma-ElectronIdentification#14__

---

1) The `LowPtGsfElectronIDProducer.cc` class has been adapted to handle changes to the list of features used by the new model in cms-data/RecoEgamma-ElectronIdentification#14

2) We have reverse engineered some recent changes made to this package in the CMSSW_11_0_X cycle. The changes involve moving functionality back to the `interface` and `src` directories. Utility methods defined in `LowPtGsfElectronFeatures.h` guarantee the correct extraction of features (from both AOD and MINIAOD data tiers) for both evaluation (in CMSSW and outside) and training purposes (using workflows defined outside the cms-sw repo).

3) These changes require the following external: cms-data/RecoEgamma-ElectronIdentification#14, which contains a "placeholder" model for the time being. The model weights will be updated again in a future PR once MC samples are available (in progress) for a final retraining. 

Validation has been performed with the "placeholder" model and the performance is consistent with our expectations.

### CODE REVIEW

**Comment by bainbrid** [comment]
@crovelli

**Comment by bainbrid** [comment]
@cavalari

**Comment by slava77** [comment]
test parameters:
- pull_request = cms-data/RecoEgamma-ElectronIdentification#14

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by bainbrid** [comment]
Hi @slava77, all:

we have yet to schedule the module to run the new ID model in this PR.

The module can run on either AOD inputs or MINIAOD inputs, transparently, but for consistency it should use the same data tier as was used for training. Currently that is MINIAOD, so we'd recommend to schedule the ID to be run on MINIAOD inputs. We achieve this ourselves using a standalone `_cfg.py` file with the following snippet:
```
process.load('RecoEgamma.EgammaElectronProducers.lowPtGsfElectronID_cfi')
process.lowPtGsfElectronID.electrons = 'slimmedLowPtElectrons'
process.lowPtGsfElectronID.rho = 'fixedGridRhoFastjetAll'
process.my_sequence *= process.lowPtGsfElectronID
```
However, we are not sure where best to add a similar snippet into the standard PAT workflows. Please could you suggest something? 
Many thanks!

**Comment by slava77** [comment]
> we have yet to schedule the module to run the new ID model in this PR.
> 
> The module can run on either AOD inputs or MINIAOD inputs, transparently

can this be added to the standard miniAOD slimmedElectrons with bParking modifier ?
@afiqaize @SohamBhattacharya please check and perhaps advise on the commands of adding the ID, in case @bainbrid doesn't already know how to add it

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc [hunk 1]
```diff
@@ -187,22 +82,39 @@ void LowPtGsfElectronIDProducer::produce(edm::StreamID, edm::Event& event, const
...
   event.getByToken(gsfElectrons_, gsfElectrons);
   if (!gsfElectrons.isValid()) {
     edm::LogError("Problem with gsfElectrons handle");
   }
 
+  // ElectronSeed unbiased BDT
+  edm::Handle<edm::ValueMap<float> > unbiasedH;
+  event.getByToken(unbiased_, unbiasedH);
+  if (!unbiasedH.isValid()) {
+    edm::LogError("Problem with unbiased handle");
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:95]: is it possible that you are using a potentially invalid handle now on line [113](https://github.com/cms-sw/cmssw/pull/31080/files#diff-bd851c5e74fefd74693080a4e490e778R113) below? Would it make sense to abort the workflow if the correct collection was not found?

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc [hunk 2]
```diff
@@ -212,16 +124,19 @@ void LowPtGsfElectronIDProducer::produce(edm::StreamID, edm::Event& event, const
     edm::ValueMap<float>::Filler filler(*ptr);
     filler.insert(gsfElectrons, output[iname].begin(), output[iname].end());
     filler.fill();
-    reco::GsfElectronRef ele(gsfElectrons, 0);
+    //reco::GsfElectronRef ele(gsfElectrons, 0);
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:127]: remove commented-out code.

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc [hunk 3]
```diff
@@ -212,16 +124,19 @@ void LowPtGsfElectronIDProducer::produce(edm::StreamID, edm::Event& event, const
...
     event.put(std::move(ptr), names_[iname]);
   }
 }
 
-double LowPtGsfElectronIDProducer::eval(const std::string& name, const reco::GsfElectronRef& ele, double rho) const {
+double LowPtGsfElectronIDProducer::eval(const std::string& name,
+                                        const edm::Ptr<reco::GsfElectron> ele,
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:133]: consider `const edm::Ptr<reco::GsfElectron>&` as good practice to avoid making a copy

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+                      maxEne2 = cluster->energy();
+                      i2 = clusNum;
+                    }
+                  }
+                  clusNum++;
+                }
+              }
+            }  // loop over clusters
+          } catch (...) {
+            std::cout << "exception caught clusNum=" << clusNum << " clus size" << sc->clustersSize()
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:287]: no `cout`, use `edm::LogError`

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+          sc_clus2_E = -999.;
+          sc_clus1_E_ov_p = -999.;
+          sc_clus2_E_ov_p = -999.;
+
+          // track-clusters match
+          clusNum = 0;
+          try {
+            if (sc->clustersSize() > 0 && sc->clustersBegin() != sc->clustersEnd()) {
+              for (auto& cluster : sc->clusters()) {
+                double pi_ = 3.1415926535;
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:307]: consider `M_PI`  rather than hardcoding the value

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+          sc_clus1_E = -999.;
+          sc_clus2_E = -999.;
+          sc_clus1_E_ov_p = -999.;
+          sc_clus2_E_ov_p = -999.;
+
+          // track-clusters match
+          clusNum = 0;
+          try {
+            if (sc->clustersSize() > 0 && sc->clustersBegin() != sc->clustersEnd()) {
+              for (auto& cluster : sc->clusters()) {
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:306]: `const auto&` should be possible.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+                  if (reach_ECAL > 0) {
+                    sc_clus2_deta = deta;
+                    sc_clus2_dphi = dphi;
+                  }
+                }
+                clusNum++;
+              }
+            }
+          } catch (...) {
+            std::cout << "caught an exception" << std::endl;
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:339]: `edm::LogError`

File: RecoEgamma/EgammaElectronProducers/interface/LowPtGsfElectronFeatures.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "DataFormats/Common/interface/Ptr.h"
+#include "DataFormats/Common/interface/Ref.h"
+#include "DataFormats/Common/interface/RefToPtr.h"
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterLazyTools.h"
+#include <vector>
+
+namespace reco {
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/interface/LowPtGsfElectronFeatures.h:12]: I think this is breaking http://cms-sw.github.io/cms_coding_rules.html#2--naming-rules-1 rule 7:

>Do not forward-declare an entity from another package.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 9]
```diff
@@ -0,0 +1,524 @@
+#include "RecoEgamma/EgammaElectronProducers/interface/LowPtGsfElectronFeatures.h"
+#include "CommonTools/BaseParticlePropagator/interface/BaseParticlePropagator.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+#include "DataFormats/ParticleFlowReco/interface/PFCluster.h"
+#include "DataFormats/ParticleFlowReco/interface/PFClusterFwd.h"
+#include "DataFormats/ParticleFlowReco/interface/PreId.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include <TVector3.h>
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:10]: I think most of the times we are including ROOT headers with `#include "TVector3.h"`, for consistency's sake.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+    float ktf_hcal_cluster_e_ = -1.;
+    float ktf_hcal_cluster_deta_ = -42.;
+    float ktf_hcal_cluster_dphi_ = -42.;
+    float preid_gsf_dpt_ = -1.;
+    float preid_trk_gsf_chiratio_ = -1.;
+    float preid_gsf_chi2red_ = -1.;
+    float trk_dxy_sig_ = -1.;  // must be last (not used by unbiased model)
+
+    // Tracks
+    reco::TrackRef trk = ecal.trackRef();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:48]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+        trk_dxy_sig_ = trk->dxyError() / trk->dxy(spot);  //@@ to be consistent with the training based on 94X MC
+      }
+      ktf_ecal_cluster_dphi_ *= trk->charge();  //@@ to be consistent with the training based on 94X MC
+    }
+
+    // Rho
+    rho_ = static_cast<float>(rho);
+
+    // ECAL clusters
+    reco::PFClusterRef ecal_clu = ecal.clusterRef();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:67]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+    if (ecal_clu.isNonnull()) {
+      ktf_ecal_cluster_e_ = ecal_clu->energy();
+      ktf_ecal_cluster_deta_ = ecal.geomMatching()[0];
+      ktf_ecal_cluster_dphi_ = ecal.geomMatching()[1];
+      ktf_ecal_cluster_e3x3_ = tools.e3x3(*ecal_clu);
+      ktf_ecal_cluster_e5x5_ = tools.e5x5(*ecal_clu);
+      auto covs = tools.localCovariances(*ecal_clu);
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:74]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+    float sc_clus1_deta = -999.;
+    float sc_clus2_deta = -999.;
+    float sc_clus1_E = -999.;
+    float sc_clus2_E = -999.;
+    float sc_clus1_E_ov_p = -999.;
+    float sc_clus2_E_ov_p = -999.;
+
+    // KF tracks
+    if (ele->core().isNonnull()) {
+      reco::TrackRef trk = ele->closestCtfTrackRef();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:171]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+        trkTV3.SetPtEtaPhi(trk->pt(), trk->eta(), trk->phi());
+        TVector3 eleTV3(0, 0, 0);
+        eleTV3.SetPtEtaPhi(ele->pt(), ele->eta(), ele->phi());
+        trk_dr = eleTV3.DeltaR(trkTV3);
+      }
+    }
+
+    // GSF tracks
+    if (ele->core().isNonnull()) {
+      reco::GsfTrackRef gsf = ele->core()->gsfTrack();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:186]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 15]
```diff
@@ -0,0 +1,524 @@
...
+      eid_brem_frac = ele->fbrem();
+      core_shFracHits = (float)ele->shFracInnerHits();
+    }
+
+    // Unbiased BDT from ElectronSeed
+    gsf_bdtout1 = unbiased;
+
+    // Clusters
+    if (ele->core().isNonnull()) {
+      reco::GsfTrackRef gsf = ele->core()->gsfTrack();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:239]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 16]
```diff
@@ -0,0 +1,524 @@
...
+      if (gsf.isNonnull()) {
+        reco::SuperClusterRef sc = ele->core()->superCluster();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:241]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 17]
```diff
@@ -0,0 +1,524 @@
...
+        gsfTV3.SetPtEtaPhi(gsf->ptMode(), gsf->etaMode(), gsf->phiMode());
+        TVector3 eleTV3(0, 0, 0);
+        eleTV3.SetPtEtaPhi(ele->pt(), ele->eta(), ele->phi());
+        gsf_dr = eleTV3.DeltaR(gsfTV3);
+      }
+    }
+
+    // Super clusters
+    if (ele->core().isNonnull()) {
+      reco::SuperClusterRef sc = ele->core()->superCluster();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:201]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 18]
```diff
@@ -0,0 +1,524 @@
...
+      }
+      if (ktf_ecal_cluster_e5x5_ > 0.) {
+        ktf_ecal_cluster_circularity_ = 1. - tools.e1x5(*ecal_clu) / ktf_ecal_cluster_e5x5_;
+      } else {
+        ktf_ecal_cluster_circularity_ = -0.1;
+      }
+    }
+
+    // HCAL clusters
+    reco::PFClusterRef hcal_clu = hcal.clusterRef();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:89]: consider `const auto&` here.

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc [hunk 19]
```diff
@@ -187,22 +82,39 @@ void LowPtGsfElectronIDProducer::produce(edm::StreamID, edm::Event& event, const
...
     output.emplace_back(gsfElectrons->size(), -999.);
   }
   for (unsigned int iele = 0; iele < gsfElectrons->size(); iele++) {
-    reco::GsfElectronRef ele(gsfElectrons, iele);
+    edm::Ptr<reco::GsfElectron> ele(gsfElectrons, iele);
+
+    if (ele->core().isNull()) {
+      continue;
+    }
+    reco::GsfTrackRef gsf = ele->core()->gsfTrack();
```
Comment (jpata) [inline RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:109]: consider `const auto&` here.

**Review by jpata — CHANGES_REQUESTED** [review]
Hi, I've checked the code and asked for a number of changes, mostly taking the chance to clean a bit the code that prepares the features.

**Comment by jpata** [comment]
If possible, can you also do a before/after timing check with `--customise Validation/Performance/TimeMemoryInfo.py`? Thanks!

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc (hunk 1 above)
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc]: As well as LogError, I now also `throw cms::Exception`. Is this "best practice" in this scenario?
I access two further handles earlier in the produce() method, so I have done the same here.

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc (hunk 2 above)
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc]: Many thanks for the comments @jpata, all have been acted on, there's just one query above. 

Regarding this:
> If possible, can you also do a before/after timing check with `--customise Validation/Performance/TimeMemoryInfo.py`? Thanks!

Does it not make more sense to wait until we have the final model ready? Essentially all the extra timing will be due to its evaluation, which will differ (perhaps substantially) from the current "placeholder" model). 

@slava77 Yes, I can sequence the module using the bParking modifier, but I'm not exactly sure of the best place to add the necessary python snippets. I can give it a go.

**Comment by bainbrid** [comment]
Many thanks for the comments @jpata, all have been acted on, there's just one query above. 

Regarding this:
> If possible, can you also do a before/after timing check with `--customise Validation/Performance/TimeMemoryInfo.py`? Thanks!

Does it not make more sense to wait until we have the final model ready? Essentially all the extra timing will be due to its evaluation, which will differ (perhaps substantially) from the current "placeholder" model). 

@slava77 Yes, I can sequence the module using the bParking modifier, but I'm not exactly sure of the best place to add the necessary python snippets. I can give it a go.

**Comment by bainbrid** [comment]
@slava77  @afiqaize @SohamBhattacharya 
[Here](https://github.com/cms-sw/cmssw/compare/14635b1..355e62b) is my attempt to add the LowPtGsfElectronIDProducer module within the PAT sequence using the bParking modifier.
Please can you take a look at the diff and tell me if it's ok.
It's tested and I will add to this PR if you're ok with it.

**Comment by jpata** [comment]
> Does it not make more sense to wait until we have the final model ready? Essentially all the extra timing will be due to its evaluation, which will differ (perhaps substantially) from the current "placeholder" model).

Fair point.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> @slava77 @afiqaize @SohamBhattacharya
> [Here](https://github.com/cms-sw/cmssw/compare/14635b1..355e62b) is my attempt to add the LowPtGsfElectronIDProducer module within the PAT sequence using the bParking modifier.
> Please can you take a look at the diff and tell me if it's ok.
> It's tested and I will add to this PR if you're ok with it.

Looking at this, I do not see where lowPtGsfElectronID is used; a module simply inserted in the task but which products are not consumed will not run.
My suggestion was to add it to the list of the electron IDs/userFloats in the slimmedElectrons.

OTOH, the main reason why I asked about this was because I thought that lowPtGsfElectronID is not already running (https://github.com/cms-sw/cmssw/pull/31080#issuecomment-671485327 says " we have yet to schedule the module to run the new ID model in this PR." ). From the reco comparisons e.g. in https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-08-06-1100+9c4a8a/38242/validateJR/all_OldVSNew_RunSinglePh2016Bwf136p731/ I see that this module is already running in default reco.

**Comment by bainbrid** [comment]
@slava77 @afiqaize @SohamBhattacharya

> Looking at this, I do not see where lowPtGsfElectronID is used; a module simply inserted in the task but which products are not consumed will not run.

Ah, true, my mistake - it is used using a standalone test we have privately, which fooled me. 

> My suggestion was to add it to the list of the electron IDs/userFloats in the slimmedElectrons.

I'm not sure how to do this. If we add a userFloat as you suggest, then everything else in this PR can remain the same? 

> OTOH, the main reason why I asked about this was because I thought that lowPtGsfElectronID is not already running ([#31080 (comment)](https://github.com/cms-sw/cmssw/pull/31080#issuecomment-671485327) says " we have yet to schedule the module to run the new ID model in this PR." ). From the reco comparisons e.g. in https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-08-06-1100+9c4a8a/38242/validateJR/all_OldVSNew_RunSinglePh2016Bwf136p731/ I see that this module is already running in default reco.

Yes, this is true, but this is part of the RECO sequence, while we wish to execute the ID as part of the PAT sequence. So the reRECO test you refer to can bre used to validate the performance within the RECO sequence (once we replace the placeholder model), but - most importantly - we wish to test the ID as part of the PAT sequence using MINIAOD inputs (i.e. same as for the training). 

Am I correct?

**Comment by jpata** [comment]
@bainbrid if you plan to run it in PAT but don't have the final model yet, I think it would be useful to already introduce the relevant python conf as suggested above to the workflows, so everything can be tested end to end.

Perhaps we can get a feeling for the timing as well, even though the model is a placeholder.

**Comment by bainbrid** [comment]
Hi @jpata. I’ve linked to an example PAT confit in a post above (sorry, can’t point directly, on phone). This is my “best guess” for an implementation. The issue is that no module then calls the ID producer (my mistake) so it’s never executed. @slava77 suggests adding to userFloats but I don’t have any experience of this. If you can point me to an example, I can attempt to add it.

**Comment by jpata** [comment]
How about consuming it here: https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/slimmedLowPtElectrons_cfi.py

**Comment by slava77** [comment]
> > My suggestion was to add it to the list of the electron IDs/userFloats in the slimmedElectrons.
> 
> I'm not sure how to do this. If we add a userFloat as you suggest, then everything else in this PR can remain the same?

The configuration logic for electron modifiers starts around 
https://github.com/cms-sw/cmssw/blob/ed643cb6e6f72d009b3f10cb01f0096aec523fdb/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L272
This would include both VID and more direct insertions in RecoEgamma/EgammaTools/python/egammaObjectModificationsInMiniAOD_cff.py
egamma experts may be more specific with a pointer.
@jainshilpi  @lsoffi  @afiqaize @SohamBhattacharya

**Comment by slava77** [comment]
> > > My suggestion was to add it to the list of the electron IDs/userFloats in the slimmedElectrons.
> > 
> > 
> > I'm not sure how to do this. If we add a userFloat as you suggest, then everything else in this PR can remain the same?
> 
> The configuration logic for electron modifiers starts around
> 
> https://github.com/cms-sw/cmssw/blob/ed643cb6e6f72d009b3f10cb01f0096aec523fdb/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L272
> 
> This would include both VID and more direct insertions in RecoEgamma/EgammaTools/python/egammaObjectModificationsInMiniAOD_cff.py
> egamma experts may be more specific with a pointer.
> @jainshilpi @lsoffi @afiqaize @SohamBhattacharya

It looks like my pointers are more specific to the standard slimmedElectrons
slimmedLowPtElectrons_cfi.py seems more appropriate in this context

**Comment by bainbrid** [comment]
@jainshilpi @lsoffi @afiqaize @SohamBhattacharya Is somebody available to take a look at this? I suspect it's straightforward for an expert to figure out how to implement this, but would take me some time to understand the MINIAOD tools. (I'm happy to implement if somebody can give pointers.) In short, our IDProducer (if called by something!) takes the slimmedLowPtElectrons collection as input and produces a ValueMap of floats, keyed off the slimmed collection. See my attempt (so far) [here](https://github.com/cms-sw/cmssw/compare/14635b1..355e62b).

**Comment by jpata** [comment]
@bainbrid the main consideration right now is that the new code is included in the jenkins tests. As has been pointed out by Slava, looks like the `LowPtGsfElectronIDProducer` is included in the RECO sequences [1] and gives some differences with respect to the baseline [2].
We can move forward with this from RECO if we understand that these differences are acceptable. Can you confirm? If the code for testing it in PAT is not immediately available, this can come later, too.

[1] https://github.com/cms-sw/cmssw/pull/31080#issuecomment-673637255
[2] https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-08-13-1100+9c4a8a/38323/validateJR/all_OldVSNew_TTbarwf25p0/all_OldVSNew_TTbarwf25p0c_floatedmValueMap_lowPtGsfElectronID__RECO_obj_values_.png

**Comment by bainbrid** [comment]
@jpata @slava77 In Slava's comment, I followed the links to [this](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-08-06-1100+9c4a8a/38242/validateJR/all_OldVSNew_RunSinglePh2016Bwf136p731/all_OldVSNew_RunSinglePh2016Bwf136p731c_floatedmValueMap_lowPtGsfElectronID__reRECO_obj_values_.png) comparison. It seems a little different to the one you point me too - perhaps just higher stats? (I will comment on this distribution.) 

I can say the following: the new distribution is shifted to lower values, which would be bad for signal and good for background. It's not clear to me if we are looking at signal or bkgd or both (presumably a mix, most likely mainly bkgd), as it depends on the MC sample used for the test. 

However, the new code we are testing here is likely to be incompatible with the model weights stored by default in cms-data, so a shift to lower values is probably my expectation. 

If you also included cms-data/RecoEgamma-ElectronIdentification#14 in this test (please can you confirm?), then the weights _would_ be compatible with the new code. For this scenario, I cannot say with certainty what I would expect to see, on balance probably _not_ a shift to lower values.  

In summary: a change is highly likely, a shift to lower values is quite possible, and the values don't seem _crazy_.

**Comment by jpata** [comment]
Thanks @bainbrid.

The test was indeed done with https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/14 as can be checked in https://github.com/cms-sw/cmssw/pull/31080#issuecomment-673541086.

The comparison Slava posted earlier in https://github.com/cms-sw/cmssw/pull/31080#issuecomment-673600480 is with `RunSinglePh2016B`, therefore higher stats for low-pt electrons.

My understanding is then:
- all the new code you added is evaluated in the tests
- it changes the output in an expected, non-final way
- the final model is in development, as are the python sequences to add it to PAT.

Did I get it right?

**Comment by jainshilpi** [comment]
> @jainshilpi @lsoffi @afiqaize @SohamBhattacharya Is somebody available to take a look at this? I suspect it's straightforward for an expert to figure out how to implement this, but would take me some time to understand the MINIAOD tools. (I'm happy to implement if somebody can give pointers.) In short, our IDProducer (if called by something!) takes the slimmedLowPtElectrons collection as input and produces a ValueMap of floats, keyed off the slimmed collection. See my attempt (so far) [here](https://github.com/cms-sw/cmssw/compare/14635b1..355e62b).



> @jainshilpi @lsoffi @afiqaize @SohamBhattacharya Is somebody available to take a look at this? I suspect it's straightforward for an expert to figure out how to implement this, but would take me some time to understand the MINIAOD tools. (I'm happy to implement if somebody can give pointers.) In short, our IDProducer (if called by something!) takes the slimmedLowPtElectrons collection as input and produces a ValueMap of floats, keyed off the slimmed collection. See my attempt (so far) [here](https://github.com/cms-sw/cmssw/compare/14635b1..355e62b).

@bainbrid apologies for missing this. We need to check with our ID and reco experts. will get back to you offline on this.

**Comment by bainbrid** [comment]
> My understanding is then:
> 
> * all the new code you added is evaluated in the tests

@jpata 

Yes.

> * it changes the output in an expected, non-final way

Yes, changes are expected. 

> * the final model is in development, as are the python sequences to add it to PAT.

Yes.

> Did I get it right?

Yes, thanks!

**Comment by jpata** [comment]
Thanks! Can I ask the EGamma contacts who were already pinged @afiqaize @SohamBhattacharya to confirm that this change to the BDT output is fine in the mean time?

**Comment by SohamBhattacharya** [comment]
Hi, sorry for missing this.
I'm checking and will get back asap.

**Comment by bainbrid** [comment]
> Thanks! Can I ask the EGamma contacts who were already pinged @afiqaize @SohamBhattacharya to confirm that this change to the BDT output is fine in the mean time?

@jpata - I'm afraid I'm probably best placed to comment on the BDT output ... but I'm afraid I cannot give a definitive answer. 

The reason in short is that the reference histogram is using a rather suboptimal model and so is a poor choice for reference (but the only one we have). Naviely, I'd expect to do better than this.

However, the model we are testing is only a placeholder and so - without quite some digging - I cannot be sure if it should give better performance than the (suboptimal) default or not. 

Further, the distributions are an (unknown) mixture of signal and background, so this complicates the interpretation. 

Finally, I'm happy to move forward. The new distribution appears reasonably healthy in that it is smooth and there are no unusual spikes (e.g. at 0 or 1). 

So green light from me. We'll know more when we have the PAT sequence and the new model.

**Comment by jpata** [comment]
Thanks @bainbrid. As you confirm it for EGamma, we can approve this PR from reco. I just want to make sure that nothing else depends on the BDT output in the mean time.

**Comment by jpata** [comment]
+reco

- tests pass with changes to the `lowPtGsfElectronID` output as expected using a placeholder model (confirmed by the author for EGamma)
- updates the `LowPtGsfElectronIDProducer` and corresponding supporting code
- model with final retraining and PAT sequences will come in a later PR

**Comment by jpata** [comment]
+1

- tests pass with changes to the lowPtGsfElectronID output as expected using a placeholder model (confirmed by the author for EGamma)
- updates the LowPtGsfElectronIDProducer and corresponding supporting code
- model with final retraining and PAT sequences will come in a later PR

**Comment by SohamBhattacharya** [comment]
Afiq and I confirm that [1] is fine as simply modifying the standard sequence to produce a new collection would be the easiest way to do this.
Since @bainbrid has already greenlit the BDT output, we're okay with these changes.

[1] https://github.com/cms-sw/cmssw/blob/355e62b7237b9e6d054c7d0fc64e1c01cc062ad4/PhysicsTools/PatAlgos/python/slimming/slimmedLowPtElectrons_cff.py

Again, sorry for the delay.

**Comment by qliphy** [comment]
+1
